File: cpp/linked_list.cpp
#include "linked_list.h"

// basic CRUD Linked list in C++
// 
// Create X
// Read X
// Update
// Destroy

LinkedList::LinkedList(string data) {
  this->head = new Node(data);
}

void LinkedList::print() {
  print(this->head);
}

void LinkedList::print(Node* currentNode) {
  cout << currentNode->data << endl;
  if (currentNode->next != NULL) print(currentNode->next);
}

void LinkedList::insert(string data) {
  insert(data, this->head);
}

void LinkedList::insert(string data, Node* currentNode) {
  if (currentNode->next == NULL) {
    currentNode->next = new Node(data);
  } else {
    insert(data, currentNode->next);
  }
}

Node* LinkedList::find(string data) {
  return find(data, this->head);
}

Node* LinkedList::find(string data, Node* currentNode) {
  if (currentNode->data == data) {
    return currentNode;
  } else {
    return find(data, currentNode->next);
  }
}

bool LinkedList::update(string toFind, string updateValue) {
  return update(toFind, updateValue, this->head);
}

bool LinkedList::update(string toFind, string updateValue, Node* currentNode) {
  Node* node = this->find(toFind);
  node->data = updateValue;

  return node->data == updateValue;
}

File: java/LinkedList.java
// Linked List basic crud - Java

// Create X
// Read X
// Update X
// Destroy X

class LinkedList {
    public Node head;

    LinkedList(String data) {
        this.head = new Node(data);
    }

    public void print() {
        print(this.head);
    }

    private void print(Node currentNode) {
        System.out.println(currentNode.data);
        if (currentNode.next != null) {
            print(currentNode.next);
        }
    }

    public void insert(String data) {
        insert(data, this.head);
    }

    private void insert(String data, Node currentNode) {
        if (currentNode.next == null) {
            currentNode.next = new Node(data);
        } else {
            insert(data, currentNode.next);
        }
    }

    public Node find(String data) {
        return find(data, this.head);
    }

    private Node find(String data, Node currentNode) {
        if (currentNode.data == data) {
            return currentNode;
        } else {
            return find(data, currentNode.next);
        }
    }

    public boolean update(String findData, String updateData) {
        return update(findData, updateData, this.head);
    }

    private boolean update(String findData, String updateData, Node currentNode) {
        if (currentNode.data == findData) {
            currentNode.data = updateData;
            return currentNode.data == updateData;
        } else if (currentNode.next != null) {
            return update(findData, updateData, currentNode.next);
        } else {
            return false;
        }
    }

    public void destroy(String data) {
        Node parent = findParent(data, this.head);
        Node node = find(data, this.head);

        if (parent == null) {
            this.head = node.next;
        } else {
            parent.next = node.next;
        }
    }

    private Node findParent(String data, Node currentNode) {
        if (this.head.data == data) {
            return null;
        } else if (currentNode.next == null) {
            return null;
        } else {
            if (currentNode.next.data == data) {
                return currentNode;
            } else {
                return findParent(data, currentNode.next);
            }
        }
    }
}
